### PR TITLE
feat(`traces`): show state changes in `cast run` and `forge test` on `-vvvvv`

### PR DESCRIPTION
**Pull request by cassc**
## Motivation

The current human-readable output from `cast run` is great for manual interpretation but requires additional parsing to be used programmatically. Additionally, there is a need to inspect storage changes for each call, but currently, these changes are not recorded.

## Solution

This PR introduces a `-j` option for `cast run` to output traces in JSON format, making it easier to consume programmatically. It also adds a `--with-state-changes` option that outputs storage changes.

### Example usage:

```bash
cast run 0xefc789b63631b255aeb6f97d948c9eac14bae4b7f021122fe24c7c6e4f34667f \
  -r https://eth.llamarpc.com -q --decode-internal --with-state-changes -j | \
  jq -R 'select(try fromjson? | . != null) | fromjson'
```

### CODE REVIEW

File: crates/verify/src/utils.rs
Comment (zerosnacks) [inline crates/verify/src/utils.rs:192]: Please undo the unrelated formatting changes

**Review by zerosnacks — CHANGES_REQUESTED** [review]
Thanks for your PR, generally in favor of adding this

File: crates/evm/traces/src/lib.rs
Comment (zerosnacks) [inline crates/evm/traces/src/lib.rs:195]: Nit:

```suggestion
    to_json: bool,
) -> String {
    if to_json {
        return serde_json::to_string(&arena.resolve_arena()).expect("Failed to write traces");
    }
    
    let mut w = TraceWriter::new(Vec::<u8>::new()).write_bytecodes(with_bytecodes);
```

File: crates/cast/bin/cmd/run.rs
Comment (zerosnacks) [inline crates/cast/bin/cmd/run.rs:61]: `--json` should just toggle the output mode

If desired it should also be added to https://github.com/paradigmxyz/revm-inspectors so the regular output mode has optional support for it as well

File: crates/verify/src/utils.rs
Comment (DaniPopes) [reply crates/verify/src/utils.rs:192]: Use nightly to format: `cargo +nightly fmt`

Comment (cassc) [reply crates/verify/src/utils.rs]: Reformatted using `cargo +nightly fmt`, thanks guys!

File: crates/cast/bin/cmd/run.rs
Comment (cassc) [reply crates/cast/bin/cmd/run.rs:61]: Totally agree. I'm happy to submit a separate PR to the `revm-inspectors` repo to add support for this.

**Comment by cassc** [comment]
Hi everyone, please let me know if there's anything else I can do to help move this PR forward.

**Comment by cassc** [comment]
For the CI error:

```rust
error: this function has too many arguments (9/7)
   --> crates/cli/src/utils/cmd.rs:357:1
    |
357 | / pub async fn handle_traces(
358 | |     mut result: TraceResult,
359 | |     config: &Config,
360 | |     chain: Option<Chain>,
...   |
366 | |     with_state_changes: bool,
367 | | ) -> Result<()> {
    | |_______________^
```

I believe we have a few options to address this:

1. Move the options `debug`, `decode_internal`, `verbose`, `json`, and `with_state_changes` into the `Config` struct.
2. Suppress the warning by adding `#[allow(clippy::too_many_arguments)]` to this function.
3. Extract the above 5 options into a new struct.

Please let me know your thoughts or suggestions.

**Comment by grandizzy** [comment]
> 2. Suppress the warning by adding #[allow(clippy::too_many_arguments)] to this function.

this should be fine IMO

File: Cargo.toml
Comment (grandizzy) [inline Cargo.toml:177]: this one was merged

Comment (cassc) [reply Cargo.toml:177]: it hasn't been included in [the lastest release](https://github.com/paradigmxyz/revm-inspectors/tags) of `revm-inspectors`, should we wait for the next release first?

Comment (cassc) [reply Cargo.toml]: hi @grandizzy I've merged the master branch, now it should have the latest `revm-inspectors` version

File: crates/cli/src/utils/cmd.rs
Comment (grandizzy) [inline crates/cli/src/utils/cmd.rs:460]: we migrated to using `sh_println!` macro for consistency but probably out of scope for this PR, @zerosnacks wdyt?

**Review by grandizzy — APPROVED** [review]
LGTM, thank you! pending other review before merging

File: crates/cli/src/utils/cmd.rs
Comment (zerosnacks) [reply crates/cli/src/utils/cmd.rs:460]: Yeah out of scope, will do a final cleanup once we enforce `sh_println!` macro usage w/ Clippy

File: crates/cast/bin/cmd/run.rs
Comment (zerosnacks) [inline crates/cast/bin/cmd/run.rs:61]:
```suggestion
    /// Prints the state changes.
    #[arg(long)]
    with_state_changes: bool,
```
This should no longer be restricted to `json` only per https://github.com/paradigmxyz/revm-inspectors/pull/213, correct?

File: crates/cli/src/utils/cmd.rs
Comment (zerosnacks) [inline crates/cli/src/utils/cmd.rs:458]: if `json` is active this `println!` should not render

This will allow directly passing to `jq` for parsing

**Review by zerosnacks — DISMISSED** [review]
Implementation looks good, some notes

File: crates/cast/bin/cmd/run.rs
Comment (cassc) [reply crates/cast/bin/cmd/run.rs]: Updated. thanks!

File: crates/cli/src/utils/cmd.rs
Comment (cassc) [reply crates/cli/src/utils/cmd.rs]: Updated, thanks! Also updated [here](https://github.com/foundry-rs/foundry/pull/9013/files#diff-448a805ef6f60ed32e529c390289b7064010d85b3da89d32d3a6bd4766723769R182) when `-q` option is used.

File: crates/cli/src/utils/cmd.rs
Comment (zerosnacks) [inline crates/cli/src/utils/cmd.rs:477]: This second check for !json isn't necessary due to the early return above

Please use the new `sh_println!`, this will make it compatible with the --quiet flag

Comment (cassc) [reply crates/cli/src/utils/cmd.rs]: Updated, thanks!

File: crates/cli/src/utils/cmd.rs
Comment (cassc) [inline crates/cli/src/utils/cmd.rs:462]: If the user is using `-j` for json output, maybe it's expected to print the output regardless of the shell verbosity settings.

**Comment by cassc** [comment]
Hi everyone, any updates on this PR? It's a bit time consuming to keep merging updates on the main branch. Thanks!

File: crates/cast/bin/cmd/run.rs
Comment (zerosnacks) [inline crates/cast/bin/cmd/run.rs:53]: We've recently added the global `shell::is_json()` to access the global `--json` flag

File: crates/cli/src/utils/cmd.rs
Comment (zerosnacks) [reply crates/cli/src/utils/cmd.rs:462]: when --json is passed some level of JSON output is to be expected where the verbosity flag `-v`, etc.. control to what degree the output is detailed

**Comment by zerosnacks** [comment]
> Any updates on this PR? It's a bit time consuming to keep merging updates on the main branch. Thank you!

Hey @cassc sorry for the delay, some small notes in regards to the global `--json` and global `-v` (verbosity) flag introduced recently. I don't think there are blockers beyond that.

File: crates/cli/src/utils/cmd.rs
Comment (cassc) [reply crates/cli/src/utils/cmd.rs:462]: does this mean in order to get the json output, we will need to pass `-v` in additonal to `-j`? I was thinking of keeping `println!` here, but it seems it has been add to `disallowed_macros` list.

**Comment by grandizzy** [comment]
@cassc could you please check https://github.com/foundry-rs/foundry/issues/2846#issuecomment-1260103355 would like to know if we could tackle all points in the PR. Thank you!

**Comment by grandizzy** [comment]
also wonder if we need the new `--with-state-changes` arg or we could just use verbosity for displaying such, @zerosnacks wdyt? thanks

**Comment by cassc** [comment]
> @cassc could you please check [#2846 (comment)](https://github.com/foundry-rs/foundry/issues/2846#issuecomment-1260103355) would like to know if we could tackle all points in the PR (or as a follow-up). Thank you!

Hi @grandizzy I think it's better to tackle in another PR, especially I think at the memoment I might not have enough knowledge about the test-level state changes.

**Comment by grandizzy** [comment]
> > @cassc could you please check [#2846 (comment)](https://github.com/foundry-rs/foundry/issues/2846#issuecomment-1260103355) would like to know if we could tackle all points in the PR (or as a follow-up). Thank you!
> 
> Hi @grandizzy I think it's better to tackle in another PR, especially I think at the memoment I might not have enough knowledge about the test-level state changes.

make sense! @mds1 could you pls chime in re UX for `cast run` point in https://github.com/foundry-rs/foundry/issues/2846#issuecomment-1260103355 ? (that is adding the `--with-state-changes` arg to cast run and the way storage diffs are displayed). Then we can extend to `forge test ` and add the cheatcode as suggested in 2nd point of your comment

**Review by zerosnacks — APPROVED** [review]
lgtm :+1:

**Comment by grandizzy** [comment]
this should be good to go but probably worth one more reviewer since @zerosnacks and I were also adding code to it

File: crates/evm/traces/src/lib.rs
Comment (mattsse) [inline crates/evm/traces/src/lib.rs:353]: one suggestion for separate improvement @zerosnacks 
let's introduce some enum for these numbers

**Review by mattsse — APPROVED** [review]
lgtm, left suggestion for @zerosnacks

File: crates/evm/traces/src/lib.rs
Comment (zerosnacks) [reply crates/evm/traces/src/lib.rs:353]: added a ticket for it here: https://github.com/foundry-rs/foundry/issues/9426

**Pull request merged**
